File: DHT11/src/DHT11_STEM.cpp
#include "DHT11_STEM.h"

DHT11_STEM::DHT11_STEM(uint8_t pin, uint8_t type)
    : _dht(pin, type), _lastTemp(0), _lastHumi(0) {}

void DHT11_STEM::begin() {
    _dht.begin();
    float temp = _dht.readTemperature();
    float hum = _dht.readHumidity();

    if (!isnan(temp)) _lastTemp = temp;
    if (!isnan(hum)) _lastHumi = hum;
}

float DHT11_STEM::readTemp() {
    float newTemp = _dht.readTemperature();
    if (isnan(newTemp)) return _lastTemp;

    float delta = newTemp - _lastTemp;
    if (delta > 0 && delta < 10) {
        _lastTemp = newTemp;
    }
    return _lastTemp;
}

float DHT11_STEM::readHumi() {
    float newHumi = _dht.readHumidity();
    if (isnan(newHumi)) return _lastHumi;

    float delta = newHumi - _lastHumi;
    if (delta > 0 && delta < 10) {
        _lastHumi = newHumi;
    }
    return _lastHumi;
}

File: ALL/DHT11_STEM.py
import dht
from machine import Pin
import math

class DHT11_STEM:
    def __init__(self, pin=32):
        self.sensor = dht.DHT11(Pin(pin))
        self._last_temp = 0
        self._last_humi = 0

    def begin(self):
        try:
            self.sensor.measure()
            self._last_temp = self.sensor.temperature()
            self._last_humi = self.sensor.humidity()
        except Exception:
            pass

    def read_temp(self):
        try:
            self.sensor.measure()
            new_temp = self.sensor.temperature()
            if not math.isnan(new_temp):
                delta = new_temp - self._last_temp
                if 0 < delta < 10:
                    self._last_temp = new_temp
        except Exception:
            pass
        return self._last_temp

    def read_humi(self):
        try:
            self.sensor.measure()
            new_humi = self.sensor.humidity()
            if not math.isnan(new_humi):
                delta = new_humi - self._last_humi
                if 0 < delta < 10:
                    self._last_humi = new_humi
        except Exception:
            pass
        return self._last_humi
